test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ClientTests/Tests/ClientEntityDescriptorTests.cs

```diff
@@ -0,0 +1,202 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ClientEntityDescriptorTests.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved.
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using Microsoft.AspNetCore.OData;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Client.E2E.TestCommon;
+using Microsoft.OData.Client.E2E.Tests.ClientTests.Server;
+using Microsoft.OData.Client.E2E.Tests.Common.Client.Default.Default;
+using Microsoft.OData.Client.E2E.Tests.Common.Server.Default;
+using Microsoft.OData.Edm;
+using Xunit;
+
+namespace Microsoft.OData.Client.E2E.Tests.ClientTests.Tests
+{
+    public class ClientEntityDescriptorTests : EndToEndTestBase<ClientEntityDescriptorTests.TestsStartup>
+    {
+        private readonly Uri _baseUri;
+        private readonly Container _context;
+        private IEdmModel _model = null;
+
+        public class TestsStartup : TestStartupBase
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.ConfigureControllers(typeof(ClientTestsController), typeof(MetadataController));
+
+                services.AddControllers().AddOData(opt => opt.Count().Filter().Expand().Select().OrderBy().SetMaxTop(null)
+                    .AddRouteComponents("odata", DefaultEdmModel.GetEdmModel()));
+            }
+        }
+
+        public ClientEntityDescriptorTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+            _context = new Container(_baseUri);
+            _model = DefaultEdmModel.GetEdmModel();
+            _context.HttpClientFactory = HttpClientFactory;
+            _context.KeyComparisonGeneratesFilterQuery = false;
+        }
+
+        private Dictionary<string, List<string>> entitySetsList = new Dictionary<string, List<string>>
+        {
+            {"People", new List<string>{"PersonID"}},
+            {"Orders", new List<string>{"OrderID"}},
+            {"ProductDetails", new List<string>{"ProductID", "ProductDetailID"}}
+        };
+
+        [Fact]
+        public void EntityDescriptor_LinksAndIdentity_AreNotNull_ForAllEntities()
+        {
+            foreach (var entitySetKeys in entitySetsList.Keys)

```

Comment:
Where does `entitySetsList` come from? Where is it defined? Also, this looks like it could have been a `[Theory]` test with some `[MemberData]` config that points to a method that returns the entity set collection instead of a `[Fact]` test with a for loop.

Reply:
`[Theory]` tests make it easier debug tests that repeat the same operation for different scenario. For example, if a test case fails, it would be easy to see for which entity set it failed from the test reports. It's also possible to run the test just for a specific entity set in Visual Studio.